### ISSUE TITLE
Run wf-human-variation on HPC cluster issue

### ISSUE THREAD

**Issue by ryao-mdanderson** [epi2me-labs/wf-human-variation#92]
### Ask away!

Dear team,

I tried to set up wf-human-variation nextflow on our institution research HPC cluster. Our login nodes have limited internet access but compute nodes do not have internet access at all. Therefore, ideally, I plan to nextflow wf-human-variation use demo file once on a login node, save the singularity images to a shared location so that all cluster users can access the  saved singularity images, furthermore, to run the pipeline on cluster compute nodes. 

I successfully ran a nextflow wf-human-variation using the demo data on login node. Next, I saved 5 images and demo data to a shared location,  try to run the nextflow pipeline on a compute node. However, I hit an error,  here is my steps to run it on a compute node:

1, module load nextflow/23.04.3

2, make a test directory :  mkdir -p $HOME/wf-human-variation-test; cd $HOME/wf-human-variation-test

3. set up environment variables
  $ export NXF_SINGULARITY_CACHEDIR="/risapps/singularity/repo/wf-human-variation/1.8.1"
     - a cluster user accessible shared location , where has 5 singularity images saved during an initial run on a login node
  $ export DEMO_DATA="/rsrch3/scratch/reflib/REFLIB_data/wf-human-variation-1.8.1/demo_data"
     - a cluster user accessible shared location for demo data
  $ OUTPUT=output

4. Run command, this is similar to what I initially do on the login node, but with two extra options '-without-docker' and '-with-singularity' (Honestly, I am not sure if just pass one image to -with-singularity is sufficient)

  $ nextflow run epi2me-labs/wf-human-variation -w ${OUTPUT}/workspace -profile singularity -without-docker -with-singularity ontresearch-wf-human-variation-sha1b503961726c6e02b6b908297a9797db953b46a3.img --snp --sv --bam $DEMO_DATA/demo.bam --bed $DEMO_DATA/demo.bed --ref $DEMO_DATA/demo.fasta --basecaller_cfg 'dna_r10.4.1_e8.2_400bps_hac@v4.1.0' --sample_name MY_SAMPLE --out_dir ${OUTPUT} 

 -- hit the error message
  ERROR ~ Error executing process > 'publish_artifact (1)'
  Caused by:
  java.io.FileNotFoundException: /risapps/singularity/repo/wf-human-variation/1.8.1/.ontresearch-wf-human-variation-sha1b503961726c6e02b6b908297a9797db953b46a3.img.img.lock (Permission denied)
   -- Check '.nextflow.log' file for details

5. Review .nextflow.log, notice the errors messages :
...
Sep-20 16:16:33.328 [main] DEBUG nextflow.script.ScriptRunner - > Awaiting termination
Sep-20 16:16:33.328 [main] DEBUG nextflow.Session - Session await
Sep-20 16:16:33.430 [Actor Thread 59] DEBUG nextflow.container.SingularityCache - Singularity found local store for image=docker://ontresearch/wf-human-variation-snp:sha0d7e7e8e8207d9d23fdf50a34ceb577da364373e; path=/risapps/singularity/repo/wf-human-variation/1.8.1/ontresearch-wf-human-variation-snp-sha0d7e7e8e8207d9d23fdf50a34ceb577da364373e.img
Sep-20 16:16:33.430 [Actor Thread 64] DEBUG nextflow.container.SingularityCache - Singularity found local store for image=docker://ontresearch/wf-human-variation-sv:shabc3ac908a14705f248cdf49f218956ec33e93ef9; path=/risapps/singularity/repo/wf-human-variation/1.8.1/ontresearch-wf-human-variation-sv-shabc3ac908a14705f248cdf49f218956ec33e93ef9.img
Sep-20 16:16:33.437 [Actor Thread 66] DEBUG nextflow.processor.TaskProcessor - Handling unexpected condition for
  task: name=bam_ingress:check_for_alignment (1); work-dir=null
  error [java.lang.IllegalStateException]: org.codehaus.groovy.runtime.InvokerInvocationException: java.io.FileNotFoundException: /risapps/singularity/repo/wf-human-variation/1.8.1/.ontresearch-wf-human-variation-sha1b503961726c6e02b6b908297a9797db953b46a3.img.img.lock (Permission denied)
Sep-20 16:16:33.437 [Actor Thread 80] DEBUG nextflow.processor.TaskProcessor - Handling unexpected condition for
  task: name=publish_artifact (1); work-dir=null
  error [java.lang.IllegalStateException]: org.codehaus.groovy.runtime.InvokerInvocationException: java.io.FileNotFoundException: /risapps/singularity/repo/wf-human-variation/1.8.1/.ontresearch-wf-human-variation-sha1b503961726c6e02b6b908297a9797db953b46a3.img.img.lock (Permission denied)
Sep-20 16:16:33.437 [Actor Thread 69] 
...

Do you have an idea what could be wrong? 

Thank you very much for your support.
Rong Yao

**Comment by cjw85** [comment]
Hi. What you've tried to do all sounds like a very sensible way of going about things given your connectivity requirements.

I'm afraid we might not be able to help a great deal and your cluster admins might better better people to ask for assistance.

The errors appear to be suggesting that singularity is having the image files your have cached. Beyond that I don't know

**Comment by ryao-mdanderson** [comment]
Hello @cjw85 :  Thank you for your quick reply. I am the cluster system admin and trying to help our cluster researchers to use nextflow pipelines.  I am stuck on the error message.

I notice a werid formatting in my last command, it should be like the followings
$  nextflow run epi2me-labs/wf-human-variation -w ${OUTPUT}/workspace -profile singularity -without-docker -with-singularity ontresearch-wf-human-variation-sha1b503961726c6e02b6b908297a9797db953b46a3.img --snp --sv --bam $DEMO_DATA/demo.bam --bed $DEMO_DATA/demo.bed --ref $DEMO_DATA/demo.fasta --basecaller_cfg 'dna_r10.4.1_e8.2_400bps_hac@v4.1.0' --sample_name MY_SAMPLE --out_dir ${OUTPUT} 
N E X T F L O W  ~  version 23.04.3

Regards,
Rong

**Comment by ryao-mdanderson** [comment]
Hello @cjw85 

We recently use the above flow (same images and reference data in a shared location) successfully to implement wf-single-cell on the cluster. It appears to work well. That is one of the reason we would like to have  wf-human-variation work in a similar way.. thinking the pipelines are developed by the same laboratory.  Anyway, thank you for the help.

**Comment by cjw85** [comment]
Yes all the wf-* workflows are developed by the same team, using standardised methods for testing, pacakaging and deployment. So if you've managed to get one workflow (with its set of container images) runnning, you should be able to get all others running. All the images are built from the same base image, only their software content is different (and all mostly installed into the images using conda).

All of that is to say I cannot conceive of a reason why wf-human-variation in particular hasn't worked for you.

@SamStudio8 is the closest thing we have to a singularity expert. He may have more suggestions.

**Comment by ryao-mdanderson** [comment]
Thank you for your reply Chris. 

I would be happy to hear from team member @SamStudio8

**Comment by ryao-mdanderson** [comment]
@RenzoTale88 
Your suggestion works! Thank you.
